lib/chef/azure/chefhandlers/exception_handler.rb

```diff
@@ -13,11 +14,19 @@ def initialize(extension_root)
 
     def report
       if run_status.failed?
+        load_run_list if not File.exists?("#{bootstrap_directory}/node-registered")
         load_azure_env
         message = "Check log file for details...\nBacktrace:\n"
         message << Array(backtrace).join("\n")
-        report_heart_beat_to_azure(AzureHeartBeat::READY, 0, "chef-service is running properly. Chef client run failed with error- #{message}") 
+        report_heart_beat_to_azure(AzureHeartBeat::READY, 0, "chef-service is running properly. Chef client run failed with error- #{message}")
       end
     end
+
+    def load_run_list
+      first_boot = File.read("#{bootstrap_directory}/first-boot.json")
+      first_boot = JSON.parse(first_boot)
+      run_list = first_boot["run_list"]
+      node.run_list.reset!(run_list)

```

Comment:
This looks good -- I was worried about adding extra communication with the Chef Server, but the idea here in general is that the Chef Server is accessible, we just want to be sure the runlist is preserved, so this is fine.